### PR TITLE
ENH, SIMD: add new universal intrinsics for trunc

### PR DESCRIPTION
**Pull request by howjmay**

### CODE REVIEW

File: numpy/core/src/common/simd/avx2/math.h [hunk 1]
```diff
@@ -109,4 +109,8 @@ NPY_FINLINE npyv_s64 npyv_min_s64(npyv_s64 a, npyv_s64 b)
 #define npyv_ceil_f32 _mm256_ceil_ps
 #define npyv_ceil_f64 _mm256_ceil_pd
 
+// trunc
+#define npyv_trunc_f32(A) _mm256_round_ps(A, (_MM_FROUND_TO_ZERO | _MM_FROUND_NO_EXC))
+#define npyv_trunc_f64(A) _mm256_round_pd(A, (_MM_FROUND_TO_ZERO | _MM_FROUND_NO_EXC))
```
Comment (seiko2plus) [inline numpy/core/src/common/simd/avx2/math.h:114]:
```suggestion
#define npyv_trunc_f32(A) _mm256_round_ps(A, _MM_FROUND_TO_ZERO)
#define npyv_trunc_f64(A) _mm256_round_pd(A, _MM_FROUND_TO_ZERO)
```
we shouldn't suppress the exceptions

File: numpy/core/tests/test_simd.py [hunk 2]
```diff
@@ -330,12 +330,15 @@ def test_square(self):
         square = self.square(vdata)
         assert square == data_square
 
-    @pytest.mark.parametrize("intrin, func", [("self.ceil", math.ceil)])
+    @pytest.mark.parametrize("intrin, func", [("self.ceil", math.ceil),
+    ("self.ceil", math.ceil)])
```
Comment (seiko2plus) [inline numpy/core/tests/test_simd.py:334]: typo I guess :)

Comment (seiko2plus) [reply numpy/core/tests/test_simd.py:334]: you must ship it to python too through [_simd.dispatch.c.src](https://github.com/numpy/numpy/blob/main/numpy/core/src/_simd/_simd.dispatch.c.src)

File: numpy/core/src/common/simd/avx2/math.h (hunk 1 above)
Comment (HowJMay) [reply numpy/core/src/common/simd/avx2/math.h]: done

File: numpy/core/tests/test_simd.py (hunk 2 above)
Comment (HowJMay) [reply numpy/core/tests/test_simd.py]: added

Comment (HowJMay) [reply numpy/core/tests/test_simd.py]: But I am meeting a weird situation now. If I understand it correctly, the addition and subtraction of the magic number 2^52 is the same as truncation. I ran the same implementation on my local machine, and I worked fine. However, the implementation became `ceil` in CI.

Comment (seiko2plus) [reply numpy/core/tests/test_simd.py]: > the addition and subtraction of the magic number 2^52 is the same as truncation

no, the idea behind it is to remove the fraction and let FPU round it according to MXCSR control mode which by default is set to "round to nearest, ties to even".

so if we assume MXCSR set to the default (Python demand it), then 
```Python
truncate = lambda x: math.copysign(abs(x) - (1.0 if round(x) > abs(x) else 0.0), x)
```

Comment (seiko2plus) [reply numpy/core/tests/test_simd.py]: according to your last commit, I think there's a miss understanding here. the example I have provided is just a pseudo-code of what you suppose to implement for `npyv_trunc_f64` on `sse2`.

Comment (HowJMay) [reply numpy/core/tests/test_simd.py]: Thank you for the information. I have fixed the implementation!

File: numpy/core/src/common/simd/sse/math.h [hunk 3]
```diff
@@ -174,4 +174,33 @@ NPY_FINLINE npyv_s64 npyv_min_s64(npyv_s64 a, npyv_s64 b)
...
+        npyv_f32 rzero = _mm_or_ps(trunc, _mm_and_ps(a, szero));
+        // if overflow return a
+        return npyv_select_f32(_mm_cmpeq_epi32(roundi, _mm_castps_si128(szero)), a, rzero);
+    }
+    NPY_FINLINE npyv_f64 npyv_trunc_f64(npyv_f64 a)
+    {
+        const __m128d szero = _mm_set1_pd(-0.0);
+        const __m128d one = _mm_set1_pd(1.0);
+        const __m128d two_power_52 = _mm_set1_pd(0x10000000000000);
+        const __m128d nonsign_bit = _mm_castsi128_pd(_mm_set1_epi64x(0x7fffffffffffffff));
```
Comment (seiko2plus) [inline numpy/core/src/common/simd/sse/math.h:197]:
```suggestion
        const __m128d nonsign_bit = _mm_castsi128_pd(npyv_setall_s64(0x7fffffffffffffffLL));
 ```
buggy on msvc(32-bit), see:
https://github.com/numpy/numpy/blob/a7e22741d058a4c1c8d6b93facbafdf712806614/numpy/core/src/common/simd/simd.h#L30-L48 

you can just use `npyv_abs_f64` instead too.

File: numpy/core/tests/test_simd.py [hunk 4]
```diff
@@ -4,6 +4,7 @@
 import itertools
 from numpy.core._simd import targets
 from numpy.core._multiarray_umath import __cpu_baseline__
+import math
```
Comment (seiko2plus) [inline numpy/core/tests/test_simd.py:7]:
```suggestion
```
already imported

**Review by seiko2plus — APPROVED** [review]
LGTM, Thank you.

**Comment by HowJMay** [comment]
> LGTM, Thank you.

Thank you @seiko2plus !

**Comment by HowJMay** [comment]
Btw I am wondering how to know if an instruction is ASIMD?
I am using this [official document website](https://developer.arm.com/architectures/instruction-sets/intrinsics/) to check the behavior of intrinsics, but I can't any clear list of ASIMD
Thank you

**Comment by mattip** [comment]
Would it be possible to add benchmarks for this and a report of any change in size of _multiarray_umath.so ?

**Comment by seiko2plus** [comment]
@HowJMay,
> Btw I am wondering how to know if an instruction is ASIMD?

well, it's kinda tricky but the short answer `neon` is only available on `armv7` and the one we use with `aarch64(arm64)` or `aarch32(backward compatible with ARMv7)` is supposed to be `ASIMD` (Advanced SIMD).

>  I am using this official document website to check the behavior of intrinsic, but I can't any clear list of ASIMD

if you saw `A32(aarch32)` or `A64(aarch64)` without `armv7` mean this intrinsic belongs to `ASIMD` only.

@mattip,
> Would it be possible to add benchmarks for this and a report of any change in size of _multiarray_umath.so ?

the same case as #19022

**Comment by mattip** [comment]
In #19022 I seem to have missed that this is now enabled for all architectures, not only FMA, AVX512F. So it should affect arm, ppc, s390x?

**Comment by seiko2plus** [comment]
#19022 and this pr too only implement universal intrinsics without using in `umath`. I have activated #19022(ceil) in #20292

> So it should affect arm, ppc, s390x?

when we use them within umath, then yes but wait we didn't support `s390x` yet. Take a look at the benchmark in #20292, and note #20292 force enable SIMD/`ceil` for `SSE` only.  please check:
https://github.com/numpy/numpy/blob/f146ec16eed3f464f152ac9be4d51e33602f4e80/numpy/core/src/umath/loops_unary_fp.dispatch.c.src#L1-L12

**Pull request merged**